LwM2M 1.10 changes: LwM2M library migrate to lwm2m_message + msg timeout callbacks

mike-scott opened the pull request:
Patch set 2 of 7(or so).

This next patchset introduces the lwm2m_message structure which will simplify sending messages as it contains all the needed information to track an LwM2M message end to end.  This is needed to add a timeout callback function for if the message is never ACK'd.  Currently, if the registration server is down or goes down during any time the client will hang and never react accordingly.  The last patch is a result of the timeout change to make sure that we don't set the lifetime too low (<15 seconds) as this is the minimum needed time for a message to expire.  If set any lower we end up sending updates faster than we can handle timeouts.

Commit:
net: lwm2m: set range for client lifetime config

Due to timeout checking the minimum lifetime must be 15 seconds,
and we're storing the lifetime as an unsigned short so set the
maximum to 65535.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: refactor engine to use lwm2m_message structure

Sending an lwm2m message is too difficult.  It requires pending / reply
and other structures to be configured and set by various portions of
the library.  There is also no way to know if a pending message ever
encounters a timeout.

Let's fix this by simplifying the internal LwM2M engine APIs for
handling lwm2m messages:

1. A user calls lwm2m_get_message(lwm2m_ctx) which returns the first
   available lwm2m message from an array of messages
   (total # of messages is set via CONFIG_LWM2M_ENGINE_MAX_MESSAGES).
2. Next the user sets all of the fields in the message that are
   required (type, code message id, token, etc)
3. Then the user calls lwm2m_init_message(msg).  This initializes the
   underlying zoap_packet, pending and reply structures.
4. Once initialized, the user creates their payload in msg->zpkt.
5. When the user is ready to send, the call lwm2m_send_message(msg).
6. And if for some reason an error occurs at any point, they can free
   up the entire set of structures with: lwm2m_release_message(msg).

Included in the refactoring is a timeout_cb field which can be set in
the LwM2M messages.  If a pending structure ever expires the engine
will call the timeout_cb passing in the msg structure before it's
automatically released.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: add timeout callbacks to registration client

Callbacks are setup for the following states:
- ENGINE_DO_BOOTSTRAP
- ENGINE_DO_REGISTRATION (first registration)
- ENGINE_REGISTRATION_DONE (subsequent client updates)
- ENGINE_DEREGISTER

In most cases, if a timeout occurs the registration engine goes back to
ENGINE_INIT.  The exception is a timeout during client update, which
forces the state machine back to ENGINE_DO_REGISTRATION (skipping a
boostrap).

Signed-off-by: Michael Scott <michael.scott@linaro.org>

rbtchc reviewed (approved):
LGTM. But shippable is complaining about the commit description.

mike-scott commented:
Woops.  Gitlint complaining about lines longer than 72 chars in the commit message in Shippable checks.  I'll re-post with a fixed commit message.

mike-scott commented:
Pushed V2 w/ gitlint fixes.  Diff: https://hastebin.com/ebihuwiles.php

mike-scott commented:
Odd.. shippable still complaining.  Researching.

mike-scott commented:
Hm.. commit description wasn't fixed in last push for some reason.  Re-doing it.

mike-scott commented:
Re-pushed with fixed commit desc for 90016fc.  Let's see what SHIPPABLE thinks.

mike-scott commented:
Ok SHIPPABLE happy now. :)

jukkar commented on include/net/lwm2m.h:
```diff
@@ -26,6 +27,36 @@
+ *
+ * @details Structure to track in-flight messages.
+ */
+struct lwm2m_message {
```
The public fields in this struct need to be documented.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -571,100 +575,190 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
+	ret = net_app_send_pkt(&msg->ctx->net_app_ctx, msg->zpkt.pkt,
+			       &msg->ctx->net_app_ctx.default_ctx->remote,
+			       NET_SOCKADDR_MAX_SIZE, K_NO_WAIT, NULL);
+	if (ret) {
```
Please have ret < 0 here for error conditions.

mike-scott commented:
posted V3 with changes recommended by @jukkar:
- Clarified the use of lwm2m_message in lwm2m.h as well as added some better comments.  It's actually a private structure used by the library but needed for the lwm2m_context since each context has it's own pool of messages.
- Check for error condition correctly (ret < 0).

Diff: https://hastebin.com/osojesuguk.cpp

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -571,100 +575,190 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
-		zoap_header_set_token(zpkt, token, tkl);
-	} else if (tkl == 0) {
-		zoap_header_set_token(zpkt, zoap_next_token(), 8);
+	if (msg->token && msg->tkl > 0) {
```
"tkl" is type u8_t and cannot be negative.

mike-scott replied:
It can be zero here though.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2354,10 +2407,59 @@ void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx, struct net_pkt *pkt,
+		msg->code = zoap_header_get_code(&response);
+		msg->mid = zoap_header_get_id(&response);
+		/* tkl -1 means don't set the token at all */
+		msg->tkl = -1;
```
tkl is type u8_t and cannot be negative.

rbtchc replied:
TKL is a 4-bit unsigned integer (per RFC 7252), we could use MSB for not generating a new token.

mike-scott replied:
or use 65535 as dont generate (what -1 equates to as a u8_t )

rbtchc replied:
Yup

mike-scott commented:
Fixing comments from @rbtchc

mike-scott commented:
Addressed comments from @rbtchc and rebased on current master.
Diff from previous: https://hastebin.com/bajarijidi.cs

mike-scott commented:
Note that this last set of changes was larger than expected, but cleans up handling token / token length variables.

mike-scott commented:
Fixed:
#define LWM2M_MSG_TOKEN_LEN_SKIP	0xFFFF
to
#define LWM2M_MSG_TOKEN_LEN_SKIP	0xFF

Which is correct for a token length of u8_t.  If we determine we need to use u16_t in the future that can be a separate patch.

rbtchc reviewed (approved):
LGTM

mike-scott commented:
@jukkar any further comments on this patchset?  I'd like to get to @rbtchc firmware support patches next.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -138,10 +140,17 @@ static char *sprint_token(const u8_t *token, u8_t tkl)
-	for (i = 0; i < tkl; i++) {
-		pos += snprintf(&buf[pos], 31 - pos, "%x", token[i]);
+	if (token && tkl != LWM2M_MSG_TOKEN_LEN_SKIP) {
+		for (i = 0; i < tkl; i++) {
```
you could move 'i' variable declaration inside that if statement.

(and aerate you code, line breaks are cheap it's nice to add an empty line after a } unless it's another } etc...)

tbursztyka commented on include/net/lwm2m.h:
```diff
@@ -26,6 +27,43 @@
+struct lwm2m_message;
+
+/* Establish a message timeout callback type */
+typedef void (*lwm2m_message_timeout_cb_t)(struct lwm2m_message *msg);
```
Is lwm2m.h included in doc generation? Because typedef are usually creating doc generation errors if not documented via \@typedef keyword etc...

tbursztyka commented on include/net/lwm2m.h:
```diff
@@ -26,6 +27,43 @@
+	/** ZoAP transmission handling structures */
+	struct zoap_pending *pending;
+	struct zoap_reply *reply;
+	u8_t send_attempts;
```
can you reorder a bit the attributes for better alignment?  (the u16_t and u8_t ones. Looks like you can gain 4 bytes I think)

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -571,100 +590,194 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
+
+	if (msg->pending) {
+		zoap_pending_clear(msg->pending);
+		msg->pending = NULL;
```
you can remove this, memset below will overwrite anyway.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -571,100 +590,194 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
+	if (msg->reply) {
+		/* make sure we want to clear the reply */
+		zoap_reply_clear(msg->reply);
+		msg->reply = NULL;
```
same

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -571,100 +590,194 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
 
-	frag = net_app_get_net_buf(app_ctx, *pkt,
+	frag = net_app_get_net_buf(app_ctx, pkt,
 				   BUF_ALLOC_TIMEOUT);
```
looks like it could fit in one line, no?

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -571,100 +590,194 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
+		zoap_header_set_token(&msg->zpkt, msg->token, msg->tkl);
+	}
+
+	if (msg->type == ZOAP_TYPE_CON) {
```
taste issue, but you could test the contrary here and return 0 if so.

Then code below would not be as indented as it is.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2227,25 +2338,16 @@ static int handle_request(struct net_app_ctx *app_ctx,
 	struct sockaddr from_addr;
-	struct zoap_packet response2;
-	struct net_pkt *pkt2;
+	struct lwm2m_message *msg = NULL;
```
in net subsystem/libs, we try to enforce declaration with assignment lines put always first. so this line would need to be below {

tbursztyka reviewed (changes requested):
tiny comments.

Commit:
net: lwm2m: Fix up lwm2m.h for doc generation

Add documentation tags for future addition to doc gen.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: cleanup token / token length handling

Instead of using a magic reference to 8 for token length, let's
establish a define for MAX_TOKEN_LENGTH and then use it for both
variable definitions and to make sure tokens are valid.  Also,

Correct the handling of a special token length value (0xFF) which
lets lwm2m_init_message() know to skip token generation.  We were
using a -1 value here previously (on a u8_t variable).

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
patchset V5 update.  Diff: https://hastebin.com/laguzupezi.cs

Notes:
- I realized I was being lazy with some documentation and token / token length changes by not splitting them out as separate patches.  So I did that here.
- [BIG CHANGE] I also realized I was making the overall addition of lwm2m_message too complicated by adding it to the lwm2m_ctx structure.  First off, it's not a public structure to be used by sample apps, etc.  And secondly the Kconfig option to set the # of maximum messages had the right idea compared to what I had actually done.  So, I removed the individual messages from lwm2m_ctx and established a single static pool in the engine.   This allowed me to move the lwm2m_message definition into the private lwm2m_engine.h file as it should have been in the first place.
- Addressed comments by @tbursztyka

mike-scott commented:
@tbursztyka @jukkar any other comments?
